pkg/server/conn.go

```diff
@@ -1139,17 +1139,23 @@ func (cc *clientConn) Run(ctx context.Context) {
 			if storeerr.ErrLockAcquireFailAndNoWaitSet.Equal(err) {
 				logutil.Logger(ctx).Debug("Expected error for FOR UPDATE NOWAIT", zap.Error(err))
 			} else {
-				var startTS uint64
-				if ctx := cc.getCtx(); ctx != nil && ctx.GetSessionVars() != nil && ctx.GetSessionVars().TxnCtx != nil {
-					startTS = ctx.GetSessionVars().TxnCtx.StartTS
+				var timestamp uint64
+				if ctx := cc.getCtx(); ctx != nil && ctx.GetSessionVars() != nil {
+					if ctx.GetSessionVars().TxnCtx != nil {
+						timestamp = ctx.GetSessionVars().TxnCtx.StartTS
+					}
+					if timestamp == 0 && ctx.GetSessionVars().StmtCtx != nil && ctx.GetSessionVars().StmtCtx.StaleReadTs > 0 {
+						// for state-read query.
+						timestamp = ctx.GetSessionVars().StmtCtx.StaleReadTs
```

Comment:
It's little weird to me to read a field of StmtCtx here, because when we reach here, the stmt should be [finished](https://github.com/pingcap/tidb/blob/master/pkg/executor/adapter.go#L1329) and related fields might be reset (like [IsStaleness](https://github.com/pingcap/tidb/blob/master/pkg/executor/adapter.go#L1382)).

How about to add a field to SessionVars (eg. LastStmtTS), then we can record it on FinishExecuteStmt and use it here (no need to resolve it from TxnCtx or StmtCtx). By this way, the logged timestamps are also consistent with those in slow logs.

Reply:
The problem is, I can't record stale-read-ts on FinishExecuteStmt, since  https://github.com/pingcap/tidb/blob/6cca89c79459911201f7ce7429ab696ed18fcc1d/pkg/session/session.go#L768

will reset the `txnManager.ctxProvider`, and then in `FinishExecuteStmt `, I can't get stale-read-ts anymore.

Reply:
Actually I don't understand why reset [IsStaleness](https://github.com/pingcap/tidb/blob/master/pkg/executor/adapter.go#L1382) here, since [IsStaleness] (https://github.com/pingcap/tidb/blob/19e61e2d35f0ca02d511637a87e590e3875832a7/pkg/executor/executor.go#L1995) will be reset in `ResetContextOfStmt`.

Reply:
> The problem is, I can't record stale-read-ts on FinishExecuteStmt

I think you've recored it in https://github.com/pingcap/tidb/pull/52484.

> I don't understand why reset IsStaleness here

The issue is not where to reset these fields, but accessing stmt scoped fields here is somehow improper.

Reply:
> The issue is not where to reset these fields, but accessing stmt scoped fields here is somehow improper.

Yes, It is a little unreasonable here. How about store `StaleReadTs` in `TxnCtx` here, it maybe more reasonable?

Reply:
> How about to add a field to SessionVars (eg. LastStmtTS), then we can record it on FinishExecuteStmt and use it here (no need to resolve it from TxnCtx or StmtCtx). By this way, the logged timestamps are also consistent with those in slow logs.

This is my proposal. "we can record it on FinishExecuteStmt" means putting `sessVars.LastStmtTS = txnTS` in `FinishExecuteStmt`. Then we can use `timestamp = ctx.GetSessionVars().LastStmtTS` here.

Reply:
I prefer to record `StaleReadTs` in `TxnCtx`. since I only care about the current statement, and this log is for current statement, so use `LastStmtTS` in this log is a little strange.

Reply:
When you reach here, no stmt is running actually.

Reply:
I already move `StaleReadTs` to `TxnCtx`, I think it's better than put it in `StmtCtx` and `SessionVar`, since `StaleReadTs` is the concept of transaction.